api/filters/free_plans.go

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2018 The Service Manager Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package filters
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/Peripli/service-manager/pkg/util"
+
+	"github.com/Peripli/service-manager/pkg/types"
+	"github.com/gofrs/uuid"
+
+	"github.com/Peripli/service-manager/pkg/log"
+
+	"github.com/Peripli/service-manager/storage"
+
+	"github.com/Peripli/service-manager/pkg/web"
+)
+
+// FreeServicePlansFilter reconciles the state of the free plans offered by all service brokers registered in SM. The
+// filter makes sure that a public visibility exists for each free plan present in SM DB.
+type FreeServicePlansFilter struct {
+	Repository storage.Repository
+}
+
+func (fsp *FreeServicePlansFilter) Name() string {
+	return "FreePlansFilter"
+}
+
+func (fsp *FreeServicePlansFilter) Run(req *web.Request, next web.Handler) (*web.Response, error) {
+	response, err := next.Handle(req)
+	if err != nil {
+		return nil, err
+	}
+	ctx := req.Context()
+	brokerID := gjson.GetBytes(response.Body, "id").String()
+	log.C(ctx).Debugf("Reconciling free plans for broker with id: %s", brokerID)
+	if err := fsp.Repository.InTransaction(ctx, func(ctx context.Context, storage storage.Warehouse) error {
+		soRepository := storage.ServiceOffering()
+		vRepository := storage.Visibility()
+
+		catalog, err := soRepository.ListWithServicePlansByBrokerID(ctx, brokerID)
+		if err != nil {
+			return err
+		}
+		for _, serviceOffering := range catalog {
+			for _, servicePlan := range serviceOffering.Plans {
+				planID := servicePlan.ID
+				isFree := servicePlan.Free
+				hasPublicVisibility := false
+				visibilitiesForPlan, err := vRepository.ListByServicePlanID(ctx, planID)
+				if err != nil {
+					return err
+				}
+				for _, visibility := range visibilitiesForPlan {
+					if isFree {
+						if visibility.PlatformID == "" {
+							hasPublicVisibility = true
+							continue
+						} else {
+							if err := vRepository.Delete(ctx, visibility.ID); err != nil {
```

Comment:
maybe a test that shows that if there are visibilities for this public plan, they are deleted and only one visibility is left?

Reply:
you cant have visibilities for paid plans directly

Its more like if the plan was paid and there were visibilities for it and then it was made public, the paid visibilities were deleted and a public visibility was added